core/primitives/src/offchain.rs

```diff
@@ -430,6 +475,147 @@ impl<T: Externalities + ?Sized> Externalities for Box<T> {
 		(&mut **self).http_response_read_body(request_id, buffer, deadline)
 	}
 }
+/// An `OffchainExternalities` implementation with limited capabilities.
+pub struct LimitedExternalities<T> {
+	capabilities: Capabilities,
+	externalities: T,
+}
+
+impl<T> LimitedExternalities<T> {
+	/// Create new externalities limited to given `capabilities`.
+	pub fn new(capabilities: Capabilities, externalities: T) -> Self {
+		Self {
+			capabilities,
+			externalities,
+		}
+	}
+
+	/// Check if given capability is allowed.
+	///
+	/// Returns `Ok` in case it is, `Err` otherwise.
+	fn check(&self, capability: Capability) -> Result<(), ()> {
+		if self.capabilities.has(capability) {
+			Ok(())
+		} else {
+			#[cfg(feature = "log")]
+			log::warn!("Accessing a forbidden API. No: {:?} capability.", capability);
+			Err(())
+		}
+	}
+}
+
+impl<T: Externalities> Externalities for LimitedExternalities<T> {
+	fn is_validator(&self) -> bool {
+		if self.check(Capability::Keystore).is_ok() {
+			self.externalities.is_validator()
+		} else {
+			false
+		}
+	}
+
+	fn submit_transaction(&mut self, ex: Vec<u8>) -> Result<(), ()> {
+		self.check(Capability::TransactionPool)?;
+		self.externalities.submit_transaction(ex)
+	}
+
+	fn network_state(&self) -> Result<OpaqueNetworkState, ()> {
+		self.check(Capability::NetworkState)?;
+		self.externalities.network_state()
+	}
+
+	fn timestamp(&mut self) -> Timestamp {
+		if self.check(Capability::Http).is_ok() {
+			self.externalities.timestamp()
+		} else {
+			Timestamp::from_unix_millis(0)
+		}
+	}
+
+	fn sleep_until(&mut self, deadline: Timestamp) {
+		if self.check(Capability::Http).is_ok() {
+			self.externalities.sleep_until(deadline)
+		}
+	}
+
+	fn random_seed(&mut self) -> [u8; 32] {
+		if self.check(Capability::Randomness).is_ok() {
+			self.externalities.random_seed()
+		} else {
+			[0_u8; 32]
+		}
```

Comment:
I think this should return a `Result`.

Reply:
Yeah, I was thinking about it. But the thing is that:
1. If you are calling it properly it will never fail.
2. If you are calling it in wrong context it will always fail.

So going into `else` branch indicates some serious issue with the runtime code - it means that someone is using these functions in places where they shouldn't. I could also consider just panicking here, but I feel this would be ugly and may potentially cause consensus issues in the future - so now I'm just printing a warning.

Reply:
@DemiMarie-parity That sounds good, but is also way more involved than simple filtering. The main reason is that runtime is a single blob that tries to import a bunch of external calls, we would need to split the runtime into multiple separate modules that import different set of external calls depending on the context (i.e. offchain workers, RuntimeApis, etc). That sounds like a serious undertake, with no huge advantages - the call would never run vs failing when run.